### PR TITLE
Optimize IN filter using bitmask

### LINKED ISSUE #13865: Optimize IN filter for integers
Currently, in Aria scan, IN filter is implemented as com.facebook.presto.orc.TupleDomainFilter.BigintValues TupleDomainFilter which builds a hash table of the IN values and performs a lookup. In case when IN list is large, but the values are relatively small, a better implementation could be to use a min, max and a bitmask. It works as follows.

- compute min and max of the values in the IN list;
- create a bitmask of max - min bits and set the bits that correspond to values in the IN list;
- test(value) method would first check if value is in [min, max] range, then check the bitmap.

The thinking is that this would be faster as long as there are < 200 bits per value. 

References: https://github.com/RoaringBitmap/RoaringBitmap

CC: @prestodb/aria

### PR DESCRIPTION
**Pull request by mbasmanova**
Fixes #13865

```
Benchmark                                (numValues)  Mode  Cnt  Score    Error  Units
BenchmarkBigintValues.lookupInBitmask           5000  avgt   20  0.020 ±  0.001  ms/op
BenchmarkBigintValues.lookupInBitmask           1000  avgt   20  0.016 ±  0.001  ms/op
BenchmarkBigintValues.lookupInBitmask            100  avgt   20  0.015 ±  0.002  ms/op
BenchmarkBigintValues.lookupInBitmask             10  avgt   20  0.017 ±  0.001  ms/op
BenchmarkBigintValues.lookupInHashTable         5000  avgt   20  0.100 ±  0.002  ms/op
BenchmarkBigintValues.lookupInHashTable         1000  avgt   20  0.087 ±  0.001  ms/op
BenchmarkBigintValues.lookupInHashTable          100  avgt   20  0.033 ±  0.001  ms/op
BenchmarkBigintValues.lookupInHashTable           10  avgt   20  0.038 ±  0.001  ms/op```
```
== NO RELEASE NOTE ==
```

### CODE REVIEW

**Comment by oerling** [comment]
It looks like the initial size should be 1 + max – min, not max – min. This does not show up as an error because BitSet seems to be self-resizing.

 

I would use a byte[] directly since almost none of the BitSet functions are used. This would save the extra range check inside BitSet.get  and the extra indirection to get to the array in the BitSet.

The difference might just be on the edge of measurable, e.g. 10 instructions.

 

To set: array[bit / 8] |= 1 << (bit & 7);

To test: flag = (array[bit / 8] & (1 << (bit & 7))) != 0;

**Comment by mbasmanova** [comment]
@oerling 

> It looks like the initial size should be 1 + max – min, not max – min. This does not show up as an error because BitSet seems to be self-resizing.

Good catch. Thank you. Fixed.

> I would use a byte[] directly since almost none of the BitSet functions are used. This would save the extra range check inside BitSet.get  and the extra indirection to get to the array in the BitSet.

I tried this, but the benchmark results got worse: 24 vs. 20 for IN list with 5000 values.

```
Benchmark                                (numValues)  Mode  Cnt  Score    Error  Units
BenchmarkBigintValues.lookupInBitmask           5000  avgt   20  0.024 ±  0.001  ms/op
BenchmarkBigintValues.lookupInBitmask           1000  avgt   20  0.019 ±  0.001  ms/op
BenchmarkBigintValues.lookupInBitmask            100  avgt   20  0.017 ±  0.002  ms/op
BenchmarkBigintValues.lookupInBitmask             10  avgt   20  0.015 ±  0.001  ms/op
BenchmarkBigintValues.lookupInHashTable         5000  avgt   20  0.101 ±  0.002  ms/op
BenchmarkBigintValues.lookupInHashTable         1000  avgt   20  0.089 ±  0.002  ms/op
BenchmarkBigintValues.lookupInHashTable          100  avgt   20  0.035 ±  0.001  ms/op
BenchmarkBigintValues.lookupInHashTable           10  avgt   20  0.038 ±  0.001  ms/op
```

BitSet class includes this comment that might explain the results above.

```
public class BitSet implements Cloneable, java.io.Serializable {
    /*
     * BitSets are packed into arrays of "words."  Currently a word is
     * a long, which consists of 64 bits, requiring 6 address bits.
     * The choice of word size is determined purely by performance concerns.
     */
    private final static int ADDRESS_BITS_PER_WORD = 6;
    private final static int BITS_PER_WORD = 1 << ADDRESS_BITS_PER_WORD;
    private final static int BIT_INDEX_MASK = BITS_PER_WORD - 1;
```

How strongly do you feel that we should re-implement necessary pieces of the BitSet ourselves?

**Comment by oerling** [comment]
This is one of those oddities one does not know till one tries. It could vary between CPU models. At this time you could just as well use BitSet. If you do, then the check against max in the caller is not needed because BitSet does this but the check against min is still needed. It could be the JVM figures out that range checks can be elided because they are done once outside and a second time inside BitSet.get. 

 

BitSet or not is not hugely important. LGTM. 

 

If this one filter were the most important piece of the workload, which it is not, we would make an encoding of 16 bit fixed width words and do a SIMD computation of the bits and indices 8 or 16 words at a time and then get a bitmask for the passing rows all in SIMD in 2 or so clocks per row. But now’s not the time.

File: presto-orc/src/main/java/com/facebook/presto/orc/TupleDomainFilterUtils.java [hunk 1]
```diff
@@ -273,4 +274,25 @@ private static TupleDomainFilter varcharRangeToFilter(Range range, boolean nullA
...
+        }
+
+        // Filter based on a hash table uses up to 3 longs per value (the value itself + 1 or 2
+        // slots in a hash table), e.g. up to 192 bits per value.
+        // Filter based on a bitmap uses (max - min) / num-values bits per value.
+        // Choose the filter that uses less bits per value.
+        if ((max - min) > Integer.MAX_VALUE || ((max - min) / values.length) > 192) {
+            return BigintValuesUsingHashTable.of(min, max, values, nullAllowed);
+        }
+        else {
```
Comment (highker) [inline presto-orc/src/main/java/com/facebook/presto/orc/TupleDomainFilterUtils.java]: redundant

File: presto-orc/src/main/java/com/facebook/presto/orc/TupleDomainFilter.java [hunk 2]
```diff
@@ -494,25 +496,30 @@ public String toString()
...
-    class BigintValues
+    class BigintValuesUsingHashTable
             extends AbstractTupleDomainFilter
     {
         private static final long EMPTY_MARKER = 0xdeadbeefbadefeedL;
         // from Murmur hash
         private static final long M = 0xc6a4a7935bd1e995L;
 
+        private final long min;
+        private final long max;
```
Comment (arhimondr) [inline presto-orc/src/main/java/com/facebook/presto/orc/TupleDomainFilter.java]: It looks like the `values` array can be removed. We can address it in a separate PR.

Comment (arhimondr) [inline presto-orc/src/main/java/com/facebook/presto/orc/TupleDomainFilter.java]: We should also add memory accounting for the arrays

File: presto-orc/src/main/java/com/facebook/presto/orc/TupleDomainFilter.java [hunk 3]
```diff
@@ -494,25 +496,30 @@ public String toString()
...
         private final int size;
         private boolean containsEmptyMarker;
 
-        private BigintValues(long[] values, boolean nullAllowed)
+        private BigintValuesUsingHashTable(long min, long max, long[] values, boolean nullAllowed)
         {
             super(true, nullAllowed);
 
             requireNonNull(values, "values is null");
+            checkArgument(min < max, "min must be less than max");
```
[the remark this answers is not in the document]
Comment (mbasmanova) [reply presto-orc/src/main/java/com/facebook/presto/orc/TupleDomainFilter.java]: Yes. We use BigintRange in that case.

File: presto-orc/src/main/java/com/facebook/presto/orc/TupleDomainFilterUtils.java [hunk 4]
```diff
@@ -105,7 +106,7 @@ public static TupleDomainFilter toFilter(Domain domain)
                     .collect(toImmutableList());
 
             if (bigintRanges.stream().allMatch(BigintRange::isSingleValue)) {
-                return BigintValues.of(
+                return toBigintValues(
```
[the remark this answers is not in the document]
Comment (mbasmanova) [reply presto-orc/src/main/java/com/facebook/presto/orc/TupleDomainFilterUtils.java]: I don't think it is necessary. TupleDomain extractor should not create duplicates. I checked some straightforward cases, e.g. in (1, 1, 1), and they work correctly.

**Comment by mbasmanova** [comment]
@highker @arhimondr James, Andrii, thank you for reviews. I updated the PR. I'll merge once Travis is green unless there are any further comments.

**Pull request merged**